internal/provisioner/kops_provisioner_cluster.go

```diff
@@ -480,41 +481,50 @@ func (provisioner *KopsProvisioner) ProvisionCluster(cluster *model.Cluster, aws
 	}
 	for daemonSet, namespace := range supportAppsWithDaemonSets {
 		if daemonSet == "k8s-spot-termination-handler" && (len(os.Getenv(model.MattermostChannel)) > 0 || len(os.Getenv(model.MattermostWebhook)) > 0) {
-			daemonSetObj, err := k8sClient.Clientset.AppsV1().DaemonSets(namespace).Get(ctx, daemonSet, metav1.GetOptions{})
-			if err != nil {
-				return errors.Wrapf(err, "failed to get daemonSet %s", daemonSet)
-			}
-
-			var payload []k8s.PatchStringValue
-			for i, envVar := range daemonSetObj.Spec.Template.Spec.Containers[0].Env {
-				if envVar.Name == "CLUSTER" {
-					payload = []k8s.PatchStringValue{{
-						Op:    "replace",
-						Path:  "/spec/template/spec/containers/0/env/" + strconv.Itoa(i) + "/value",
-						Value: cluster.ID,
-					}}
+			attempts := 5
+			var daemonSetObj *appsv1.DaemonSet
+			for attempt := 0; attempt < attempts; attempt++ {
+				daemonSetObj, err = k8sClient.Clientset.AppsV1().DaemonSets(namespace).Get(ctx, daemonSet, metav1.GetOptions{})
+				if err != nil && attempt < 5 {
+					time.Sleep(5 * time.Second)
+					continue
```

Comment:
I think the issue is actually here https://github.com/mattermost/mattermost-cloud/pull/624/files#diff-c47cc93d0609e79ef543c8fb97a4e2e1c222380084d5340f6cd8441e75b7fad7R269-R271

We initialize context with a timeout of 61 seconds, we shadow this variable in couple places, but here we probably refer to the context with the 61 seconds timeout, therefore if all other things will not complete in a given time the query will fail with `context deadline exceeded` as we observed it.

I would suggest making new context for each Kubernetes API based on the context we want to have for whole provisioning, which probably should be more than 61 seconds. I think that should solve the issue if not we can proceed with retries.

Reply:
@Szymongib so to clarify your proposal is to create new context for this or to increase the wait time to 120 f.e?

Reply:
Yeah, so we can start by increasing the wait time, we might need to cleanup/standardize how we handle context there.